File: image_filtering/task.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

void bluredImage(cv::Mat& input, cv::Mat& output);
void sharpedFilter(cv::Mat& input, cv::Mat& output);
void medianBlurFilter(cv::Mat& input, cv::Mat& output);
bool save_image(cv::Mat& image, const std::string& fileName);

int main()
{
    std::string fileLAB = "fileLab.png";
    std::string fileHSV = "fileHSV.png";
    std::string fileRGB = "fileRGB.png";
    std::string fileGRAY = "fileGRAY.png";
    std::string fileBlur = "fileblur.png";
    std::string fileMedian = "filemedian.png";
    std::string fileSharped = "filesharped.png";
    cv::Mat image = cv::imread("image.jpg", -1);
    cv::Mat imageLAB, imageRGB, imageHSV, imageGRAY, blurImage, medianImage, sharpedImage;
    if (image.empty()) {
        std::cout << "Image is empty" << std::endl;
        return -1;
    }

    cv::cvtColor(image, imageHSV, cv::COLOR_BGR2HSV);
    cv::cvtColor(image, imageLAB, cv::COLOR_BGR2Lab);
    cv::cvtColor(image, imageRGB, cv::COLOR_BGR2RGB);
    cv::cvtColor(image, imageGRAY, cv::COLOR_BGR2GRAY);

    cv::imshow("HSV space", imageHSV);
    cv::imshow("LAB space", imageLAB);
    cv::imshow("RGB space", imageRGB);
    cv::imshow("GRAY space", imageGRAY);
    
    bluredImage(image, blurImage);
    cv::imshow("Gaussian blur", blurImage);
    sharpedFilter(image, sharpedImage);
    cv::imshow("Sharped filter", sharpedImage);
    medianBlurFilter(image, medianImage);
    cv::imshow("Median filter", medianImage);

    save_image(imageHSV, fileHSV);
    save_image(imageLAB, fileLAB);
    save_image(imageRGB, fileRGB);
    save_image(imageGRAY, fileGRAY);
    save_image(blurImage, fileBlur);
    save_image(sharpedImage, fileSharped);
    save_image(medianImage, fileMedian);
    cv::waitKey(0);
}

void bluredImage(cv::Mat& input, cv::Mat& output)
{
    cv::Size kernel(5, 5);
    double x = 0;
    double y = 0;
    cv::GaussianBlur(input, output, kernel, x, y);
}

void sharpedFilter(cv::Mat& input, cv::Mat& output)
{
    cv::Mat kernel = (cv::Mat_<float>(3, 3) <<
        0, -1, 0,
        -1, 5, -1,
        0, -1, 0);
    cv::Mat sharpenedImage;
    cv::filter2D(input, output, input.depth(), kernel);
}

void medianBlurFilter(cv::Mat& input, cv::Mat& output)
{
    int kernel = 15;
    cv::medianBlur(input, output, kernel);
}

bool save_image(cv::Mat& image, const std::string& fileName)
{
    if (image.empty()) {
        std::cerr << "Image is empty failed Saving" << std::endl;
        return false;
    }

    try {
        cv::imwrite(fileName, image);
        std::cout << "Image saved successfuly" << std::endl;
        return true;
    } catch(const cv::Exception& signal) {
        std::cerr << "Error: " << signal.what() << std::endl;
        return false;
    }
}

File: image_filtering/task.py
import cv2

def blured_image(input_image):
    kernel_size = (5, 5)
    blurred_image = cv2.GaussianBlur(input_image, kernel_size, 0)
    return blurred_image

def sharped_filter(input_image):
    kernel = np.array([
        [0, -1, 0],
        [-1, 5, -1],
        [0, -1, 0]
    ], dtype=np.float32)
    sharpened_image = cv2.filter2D(input_image, -1, kernel)
    return sharpened_image

def median_blur_filter(input_image):
    kernel_size = 15
    median_blurred_image = cv2.medianBlur(input_image, kernel_size)
    return median_blurred_image

def save_image(image, file_name):
    try:
        cv2.imwrite(file_name, image)
        print("Image saved successfully")
        return True
    except cv2.error as e:
        print(f"Error: {e}")
        return False

if __name__ == "__main__":
    file_lab = "fileLab.png"
    file_hsv = "fileHSV.png"
    file_rgb = "fileRGB.png"
    file_gray = "fileGRAY.png"
    file_blur = "fileblur.png"
    file_median = "filemedian.png"
    file_sharped = "filesharped.png"

    image = cv2.imread("image.jpg")
    if image is None:
        print("Image is empty")
        exit(1)

    image_lab = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
    image_hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image_gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    cv2.imshow("HSV space", image_hsv)
    cv2.imshow("LAB space", image_lab)
    cv2.imshow("RGB space", image_rgb)
    cv2.imshow("GRAY space", image_gray)

    blur_image = blured_image(image)
    cv2.imshow("Gaussian blur", blur_image)

    sharpened_image = sharped_filter(image)
    cv2.imshow("Sharpened filter", sharpened_image)

    median_image = median_blur_filter(image)
    cv2.imshow("Median filter", median_image)

    save_image(image_hsv, file_hsv)
    save_image(image_lab, file_lab)
    save_image(image_rgb, file_rgb)
    save_image(image_gray, file_gray)
    save_image(blur_image, file_blur)
    save_image(sharpened_image, file_sharped)
    save_image(median_image, file_median)

    cv2.waitKey(0)
    cv2.destroyAllWindows()

